### PR TITLE
test[next]: simplify boilerplate for fieldview single feature tests

### PR DESCRIPTION
**Pull request by DropD**
## Description

Added:
- `next_tests.feature_tests.cases`
  - helpers to express repetitive test setup concisely

Changed:
- `next_tests.feature_tests.ffront_tests.test_execution`
  - several test methods use the "cases" helpers
 
## Requirements

- [X] All fixes and/or new features come with corresponding tests.
- [x] Important design decisions have been documented in the approriate ADR inside the [docs/development/ADRs/](docs/development/ADRs/Index.md) folder.

### CODE REVIEW

File: tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py [hunk 1]
```diff
@@ -38,296 +38,255 @@
 from gt4py.next.iterator.embedded import index_field, np_as_located_field
 from gt4py.next.program_processors.runners import gtfn_cpu
 
+from next_tests.integration_tests.feature_tests import cases
+from next_tests.integration_tests.feature_tests.cases import (
```
Comment (havogt) [inline tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py:42]: I would violate our style guides even more, but I am probably alone
`from next_tests.integration_tests.feature_tests.cases import *`

File: tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py [hunk 2]
```diff
@@ -38,296 +38,255 @@
...
+    @field_operator
+    def shift_by_one(inp: cases.VField) -> cases.EField:
         return inp(E2V[0])
 
-    shift_by_one(a, out=b, offset_provider={"E2V": reduction_setup.offset_provider["E2V"]})
+    inp = cases.allocate(unstructured_case, shift_by_one, "inp")()
+    out = cases.allocate(unstructured_case, shift_by_one, cases.RETURN)()
 
-    ref = np.asarray(a)[reduction_setup.offset_provider["E2V"].table[slice(0, None), 0]]
+    ref = np.asarray(inp)[unstructured_case.offset_provider["E2V"].table[slice(0, None), 0]]
```
Comment (havogt) [inline tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py:102]:
```suggestion
    ref = inp[unstructured_case.offset_provider["E2V"].table[:, 0]]
```
does this work?

File: tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py [hunk 3]
```diff
@@ -38,296 +38,255 @@
...
     """Shifting the result of an addition should work."""
-    a = np_as_located_field(IDim)(np.arange(size + 1, dtype=np.float64))
-    b = np_as_located_field(IDim)(np.ones((size + 2)) * 2)
-    out_I_float = np_as_located_field(IDim)(np.zeros((size,), dtype=float64))
 
     @field_operator
-    def auto_lift(
-        inp1: Field[[IDim], float64], inp2: Field[[IDim], float64]
-    ) -> Field[[IDim], float64]:
+    def auto_lift(inp1: cases.IJKField, inp2: cases.IJKField) -> cases.IJKField:
```
Comment (havogt) [inline tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py:110]: If test is already described by the name of the test function, I'd use a generic name in the field_operator, it seems we are often copy-pasting anyway.
```suggestion
    def testee(inp1: cases.IJKField, inp2: cases.IJKField) -> cases.IJKField:
```

File: tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py [hunk 4]
```diff
@@ -38,296 +38,255 @@
...
         return tmp(Ioff[1])
 
-    @program(backend=fieldview_backend)
-    def fencil(
-        inp1: Field[[IDim], float64], inp2: Field[[IDim], float64], out: Field[[IDim], float64]
-    ) -> None:
-        auto_lift(inp1, inp2, out=out)
+    a = cases.allocate(cartesian_case, auto_lift, "inp1").extend({cases.IDim: (0, 1)})()
+    b = cases.allocate(cartesian_case, auto_lift, "inp2").extend({cases.IDim: (0, 2)})()
+    out_I_float = cases.allocate(cartesian_case, auto_lift, cases.RETURN)()
```
Comment (havogt) [inline tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py:116]:
```suggestion
    a = cases.allocate(cartesian_case, testee, "inp1").extend({cases.IDim: (0, 1)})()
    b = cases.allocate(cartesian_case, testee, "inp2").extend({cases.IDim: (0, 2)})()
    out = cases.allocate(cartesian_case, testee, cases.RETURN)()
```

File: tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py [hunk 5]
```diff
@@ -38,296 +38,255 @@
...
 
-    fencil(a, b, out_I_float, offset_provider={"Ioff": IDim})
+    cases.verify(cartesian_case, auto_lift, a, b, out=out_I_float, ref=a[1:] + b[2:])
```
Comment (havogt) [inline tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py:118]:
```suggestion
    cases.verify(cartesian_case, testee, a, b, out=out, ref=a[1:] + b[2:])
```

File: tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py [hunk 6]
```diff
@@ -38,296 +38,255 @@
...
-    inp = 5.0
-    out = np_as_located_field(Vertex)(np.zeros([size]))
+    inp = cases.allocate(unstructured_case, scalar_arg, "scalar_inp")()
+    out = cases.allocate(unstructured_case, scalar_arg, cases.RETURN)()
+    ref = np.full([unstructured_case.default_sizes[Vertex]], inp + 1)
 
-    @field_operator(backend=fieldview_backend)
-    def scalar_arg_inner(scalar_inp: float64) -> Field[[Vertex], float64]:
-        return broadcast(scalar_inp + 1.0, (Vertex,))
+    # todo(ricoh): replace with verify_with_default_data once grid type can be set on fielops
```
Comment (havogt) [inline tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py:144]: this was merged, so we an already use it here

File: tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py [hunk 7]
```diff
@@ -38,296 +38,255 @@
...
+        return scalar_arg_inner(scalar_inp + 1)
 
+    inp = cases.allocate(unstructured_case, scalar_arg, "scalar_inp")()
+    out = cases.allocate(unstructured_case, scalar_arg, cases.RETURN)()
+    ref = np.full([unstructured_case.default_sizes[Vertex]], inp + 2)
 
-def test_scalar_arg_with_field(fieldview_backend):
-    if fieldview_backend in [gtfn_cpu.run_gtfn, gtfn_cpu.run_gtfn_imperative]:
-        pytest.skip("IndexFields and ConstantFields are not supported yet.")
+    # todo(ricoh): replace with verify_with_default_data once grid type can be set on fielops
```
Comment (havogt) [inline tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py:163]: see above

File: tests/next_tests/integration_tests/feature_tests/cases.py [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+
+RETURN = "return"
+
+
+def allocate_(
+    case: Case,
+    fieldview_prog: decorator.FieldOperator | decorator.Program,
+    name: str,
+    sizes: dict[common.Dimension, int],
+    strategy: Optional[FieldInitializer | ScalarInitializer] = None,
```
Comment (havogt) [inline tests/next_tests/integration_tests/feature_tests/cases.py:186]: If `strategy` only refers to initialization strategy, I'd call it differently. Maybe `init` or `initializer`.
If the `strategy` function will be used often, I'd also consider promoting some initializers to the top level, such that
```
qc = cases.allocate(cartesian_case, scan_tuple_scalar, "qc").strategy(cases.zeros)()
```
becomes
```
qc = cases.allocate(cartesian_case, scan_tuple_scalar, "qc").zeros()()
```

File: tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py [hunk 9]
```diff
@@ -38,296 +38,255 @@
...
+    cases.verify(cartesian_case, scan_tuple_scalar, qc, tuple_scalar, out=qc, ref=expected)
 
-def test_astype_int(fieldview_backend):
-    size = 10
-    b_float_64 = np_as_located_field(IDim)(np.full((size,), fill_value=1.5, dtype=np.float64))
-    c_int64 = np_as_located_field(IDim)(np.ones((size,), dtype=np.int64))
-    out_int_64 = np_as_located_field(IDim)(np.zeros((size,), dtype=np.int64))
 
-    @field_operator(backend=fieldview_backend)
+def test_astype_int(cartesian_case):  # noqa: F811 # fixtures
```
Comment (havogt) [inline tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py:250]: I think we should put the 3 following tests into a parametrized one. Could you try if that's compatible with the new infrastructure? (Disclaimer: not sure if its easily doable without the new infrastructure.

**Review by havogt — CHANGES_REQUESTED** [review]

File: tests/next_tests/integration_tests/feature_tests/cases.py [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+    return embedded.np_as_located_field(*sizes.keys())(
+        np.arange(np.prod(svals), dtype=dtype).reshape(svals)
+    )
+
+
+def scalar5(dtype: str, shape: Optional[tuple[int, ...]]) -> int | float | np.array:
+    """Initialize a scalar to the value 5."""
+    if shape:
+        return np.ones(np.prod(shape), dtype=dtype) * 5
+    return np.dtype(dtype).type(5)
```
[the remark this answers is not in the document]
Comment (DropD) [reply tests/next_tests/integration_tests/feature_tests/cases.py:96]: Why not?

File: tests/next_tests/integration_tests/feature_tests/cases.py [hunk 11]
```diff
@@ -0,0 +1,358 @@
+# GT4Py - GridTools Framework
```
[the remark this answers is not in the document]
Comment (DropD) [reply tests/next_tests/integration_tests/feature_tests/cases.py:1]: I am not set on it.

File: tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py (hunk 1 above)
Comment (DropD) [reply tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py:42]: I am willing to do this only if the style guide is changed (could allow for this in tests or other well defined circumstances).

File: tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py (hunk 9 above)
Comment (DropD) [reply tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py:250]: I don't currently see a way to do it without extending the toolchain. `astype(x, <type_builtin>)` does not look into `<type_builtin>` to see if it is a variable containing a type builtin. The toolchain only looks at the name.

File: tests/next_tests/integration_tests/feature_tests/cases.py [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+    extend: Optional[dict[common.Dimension, tuple[int, int]]] = None,
+) -> dict[common.Dimension, str]:
+    sizes = sizes.copy()
+    if extend:
+        for dim, (lower, upper) in extend.items():
+            sizes[dim] += upper - lower
+    return sizes
+
+
+RETURN = "return"
```
Comment (egparedes) [inline tests/next_tests/integration_tests/feature_tests/cases.py:178]:
```suggestion
#: <Add some description here>
RETURN = "return"
```

File: tests/next_tests/integration_tests/feature_tests/cases.py [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+        raise ValueError(
+            f"test cases do not support {type(fieldview_prog)} with empty .definition attribute (as you would get from .as_program())!"
+        )
+    annotations = typing.get_type_hints(fieldview_prog.definition)
+    return {
+        name: type_translation.from_type_hint(type_hint) for name, type_hint in annotations.items()
+    }
+
+
+def extend_sizes(
```
Comment (egparedes) [inline tests/next_tests/integration_tests/feature_tests/cases.py:167]: Missing docstring.

File: tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py (hunk 9 above)
Comment (DropD) [reply tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py:250]: Everything else should work though:
 ✅  field type annotations parametrized by dtype
 ✅  allocating the output fields parametrized by dtype
 ✅  reference parametrized by dtype
 ❌  field operator with `astype` parametrized by dtype

File: tests/next_tests/integration_tests/feature_tests/cases.py [hunk 14]
```diff
@@ -0,0 +1,358 @@
...
+    if extend:
+        for dim, (lower, upper) in extend.items():
+            sizes[dim] += upper - lower
+    return sizes
+
+
+RETURN = "return"
+
+
+def allocate_(
```
Comment (egparedes) [inline tests/next_tests/integration_tests/feature_tests/cases.py:181]: Just a question: why not `_allocate()`? If the builder is meant to be the only interface this one should be hidden, right? Or if both are needed for some corner cases, maybe we could use something more explicit like `allocate_field`?

File: tests/next_tests/integration_tests/feature_tests/cases.py [hunk 15]
```diff
@@ -0,0 +1,358 @@
...
+        comparison=comparison,
+    )
+
+
+@dataclasses.dataclass
+class Case:
+    """Parametrizable components for single feature integration tests."""
+
+    backend: ppi.ProgramProcessor
+    offset_provider: dict[str, common.Connectivty | common.Dimension]
```
Comment (egparedes) [inline tests/next_tests/integration_tests/feature_tests/cases.py:325]:
```suggestion
    offset_provider: dict[str, common.Connectivity | common.Dimension]
```

File: tests/next_tests/integration_tests/feature_tests/cases.py [hunk 16]
```diff
@@ -0,0 +1,358 @@
...
+
+def allocate_(
+    case: Case,
+    fieldview_prog: decorator.FieldOperator | decorator.Program,
+    name: str,
+    sizes: dict[common.Dimension, int],
+    strategy: Optional[FieldInitializer | ScalarInitializer] = None,
+    dtype: Optional[str] = None,
+    extend: Optional[dict[common.Dimension, tuple[int, int]]] = None,
+) -> common.Field | tuple[common.Field, ...] | int | float:
```
Comment (egparedes) [inline tests/next_tests/integration_tests/feature_tests/cases.py:189]: For coherency with previous suggestions about scalar value types:
```suggestion
) -> common.Field | tuple[common.Field, ...] | ScalarValue:
```

File: tests/next_tests/integration_tests/feature_tests/cases.py [hunk 17]
```diff
@@ -0,0 +1,358 @@
...
+
+def verify(
+    case: Case,
+    fieldview_prog: decorator.FieldOperator | decorator.Program,
+    *args: common.Field,
+    out: Optional[common.Field] = None,
+    nopass_out: Optional[common.Field] = None,
+    ref: common.Field,
+    offset_provider: Optional[dict[str, common.Connectivty | common.Dimension]] = None,
+    comparison: Callable[[Any, Any], bool] = np.allclose,
```
Comment (egparedes) [inline tests/next_tests/integration_tests/feature_tests/cases.py:272]:
```suggestion
    ref: common.Field,
    out: Optional[common.Field] = None,
    nopass_out: Optional[common.Field] = None,
    offset_provider: Optional[dict[str, common.Connectivty | common.Dimension]] = None,
    comparison: Callable[[Any, Any], bool] = np.allclose,
```

File: tests/next_tests/integration_tests/feature_tests/cases.py [hunk 18]
```diff
@@ -0,0 +1,358 @@
...
+            case,
+            fieldview_prog,
+            *args,
+            out=out,
+            offset_provider=offset_provider,
+        )
+    else:
+        run(case, fieldview_prog, *args, offset_provider=offset_provider)
+
+    assert comparison(ref, out or nopass_out)
```
Comment (egparedes) [inline tests/next_tests/integration_tests/feature_tests/cases.py:292]: I don't understand how the `nopass_out` argument works here...

File: tests/next_tests/integration_tests/feature_tests/cases.py [hunk 19]
```diff
@@ -0,0 +1,358 @@
...
+
+
+def verify_with_default_data(
+    case: Case,
+    fieldop: decorator.FieldOperator,
+    ref: Callable,
+    comparison: Callable[[Any, Any], bool] = np.allclose,
+) -> None:
+    """
+    Check the result of executing a fieldview code object agains a reference constructed from the inputs.
```
Comment (egparedes) [inline tests/next_tests/integration_tests/feature_tests/cases.py:302]:
```suggestion
    Check the result of executing a fieldview code object against a reference constructed from the inputs.
```

File: tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py (hunk 1 above)
Comment (egparedes) [reply tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py:42]: I disagree. I don't think star imports are a good idea except in quick interactive prototyping sessions.

File: tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py (hunk 3 above)
Comment (egparedes) [reply tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py:110]: Agree

File: tests/next_tests/integration_tests/feature_tests/cases.py [hunk 20]
```diff
@@ -0,0 +1,358 @@
...
+@dataclasses.dataclass(frozen=True)
+class Builder:
+    partial: functools.partial
+
+    def build(self, *args, **kwargs) -> Self:
+        return self.partial(*args, **kwargs)
+
+    def __call__(self, *args, **kwargs) -> Any:
+        return self.build(*args, **kwargs)
+
```
Comment (egparedes) [inline tests/next_tests/integration_tests/feature_tests/cases.py:108]: I think a custom `__getattr__()` should be added here  (maybe raising a custom message error) to avoid complaints from the static type checker, since the attribute setters are added at runtime.

**Review by egparedes — CHANGES_REQUESTED** [review]
I have some comments and questions.

File: tests/next_tests/integration_tests/feature_tests/cases.py (hunk 12 above)
Comment (tehrengruber) [reply tests/next_tests/integration_tests/feature_tests/cases.py:178]: Is a string required here? If not I would just use some unique sentinel value. Otherwise you could use just use "out" but as a convention.
```suggestion
RETURN = "out"
```

**Review by tehrengruber — CHANGES_REQUESTED** [review]
Partial review only.

File: tests/next_tests/integration_tests/feature_tests/cases.py [hunk 21]
```diff
@@ -0,0 +1,358 @@
...
+        .case(case)
+        .fieldview_prog(fieldview_prog)
+        .name(name)
+    )
+
+
+def get_param_types(
+    fieldview_prog: decorator.FieldOperator | decorator.Program,
+) -> dict[str, ts.FieldType | ts.TupleType | ts.ScalarType]:
+    if fieldview_prog.definition is None:
```
[the remark this answers is not in the document]
Comment (DropD) [reply tests/next_tests/integration_tests/feature_tests/cases.py:269]: So what you said about reading from annotations being more robust was only for the return type?

[the remark this answers is not in the document]
Comment (DropD) [reply tests/next_tests/integration_tests/feature_tests/cases.py:269]: Does not work with the current interface: `....type.definition` doesn't contain the names of args that aren't keyword-only. Should we change the interface to allocate by index of the argument instead of name? Alternatively we could mandate kwonly args in tests?

File: tests/next_tests/integration_tests/feature_tests/cases.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (DropD) [reply tests/next_tests/integration_tests/feature_tests/cases.py:1]: It is already in the `next_tests` package, so `utils` should suffice.

[the remark this answers is not in the document]
Comment (DropD) [reply tests/next_tests/integration_tests/feature_tests/cases.py:1]: I agree and would like to raise the following question:

If both of the following are true:

1. We all agree this is useful for learning / prototyping sessions
2. We all agree it should be tested

Does it not follow:

3. (1 + 2) It should be part of the library?

File: tests/next_tests/integration_tests/feature_tests/cases.py (hunk 18 above)
Comment (DropD) [reply tests/next_tests/integration_tests/feature_tests/cases.py]: I renamed it and documented it (hopefully) more specifically in the docstring. Is it more understandable now?

File: tests/next_tests/integration_tests/feature_tests/cases.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (DropD) [reply tests/next_tests/integration_tests/feature_tests/cases.py:1]: The consensus seems to be that parts of this should be moved to the library in a later PR. This also means we will hold off on merging this with other test utils for now.

[the remark this answers is not in the document]
Comment (DropD) [reply tests/next_tests/integration_tests/feature_tests/cases.py:1]: Holding off on merging to simplify moving some of these utils into gt4py.next library.

File: tests/next_tests/integration_tests/feature_tests/cases.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (tehrengruber) [reply tests/next_tests/integration_tests/feature_tests/cases.py:284]: We discussed this and concluded to postpone any action here.

File: tests/next_tests/integration_tests/feature_tests/cases.py [hunk 22]
```diff
@@ -0,0 +1,358 @@
...
+    fieldview_backend,
+)
+
+
+IField: TypeAlias = common.Field[[IDim], int]
+IJKField: TypeAlias = common.Field[[IDim, JDim, KDim], int]
+IJKFloatField: TypeAlias = common.Field[[IDim, JDim, KDim], np.float64]
+VField: TypeAlias = common.Field[[Vertex], int]
+EField: TypeAlias = common.Field[[Edge], int]
+V2EDim = common.Dimension("V2E", kind=common.DimensionKind.LOCAL)
```
[the remark this answers is not in the document]
Comment (tehrengruber) [reply tests/next_tests/integration_tests/feature_tests/cases.py:55]: We discussed this and concluded to avoid changing anything in `reduction_setup` for now. @DropD Please add the todo here that the duplication of `V2EDim`, `E2VDim`, `V2E`, `E2V` should be removed.

File: tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py [hunk 23]
```diff
@@ -149,223 +141,189 @@ def composed_shift_unstructured(inp: Field[[Vertex], float64]) -> Field[[Cell],
...
+    @field_operator
+    def testee(a: Field[[IDim], float64]) -> Field[[IDim], np.float32]:
+        b = astype(a, float32)
+        return b
+
+    cases.verify_with_default_data(
+        cartesian_case,
+        testee,
+        ref=lambda a: a.astype(np.float32),
+        comparison=lambda a, b: np.all(np.equal(a, b)),
```
Comment (tehrengruber) [inline tests/next_tests/integration_tests/feature_tests/ffront_tests/test_execution.py:325]:
```suggestion
        comparison=lambda a, b: np.all(a == b),
```
As used elsewhere.

**Review by tehrengruber — APPROVED** [review]
Only small things left.

**Review by egparedes — APPROVED** [review]
It looks good to me. I have some (optional) suggestions and a couple of small docstring and typing fixes, but I don't need a new review cycle for them so I approve the PR now.

File: tests/next_tests/integration_tests/feature_tests/cases.py [hunk 24]
```diff
@@ -0,0 +1,570 @@
...
+
+
+@typing.overload
+def make_builder(
+    *args: Optional[Callable], **kwargs: dict[str, Any]
+) -> Callable[[Callable], Callable[..., Builder]] | Callable[..., Builder]:
+    ...
+
+
+def make_builder(
```
[the remark this answers is not in the document]
Comment (DropD) [reply tests/next_tests/integration_tests/feature_tests/cases.py:221]: I will ad a todo. It would probably take some experimentation to get the most out of it.

File: tests/next_tests/integration_tests/feature_tests/cases.py [hunk 25]
```diff
@@ -0,0 +1,570 @@
...
+
+    def field(
+        self,
+        backend: ppi.ProgramProcessor,
+        sizes: dict[common.Dimension, int],
+        dtype: np.typing.DTypeLike,
+    ) -> FieldValue:
+        return embedded.np_as_located_field(*sizes.keys())(
+            np.zeros(tuple(sizes.values()), dtype=dtype)
+        )
```
[the remark this answers is not in the document]
Comment (DropD) [reply tests/next_tests/integration_tests/feature_tests/cases.py]: Sure. It would break if `ConstInitializer` was made frozen, but I have decided that I am not going to at this point.

**Review by havogt — APPROVED** [review]
I didn't do another round of review, but it already got a lot of feedback. Let's merge.
